Add pinned memory support to sparse COO/CSR/CSC/BSR/BSC tensors

pearu opened the pull request:
As in the title:

To register indices/values of a sparse XYZ tensor with CUDA, the following methods are supported
- `sparse_xyz_tensor(indices, values, pin_memory=True)`
- `sparse_xyz_tensor(indices, values).pin_memory()`
- `sparse_xyz_tensor(indices.pin_memory(), values.pin_memory())`

Fixes https://github.com/pytorch/pytorch/issues/115330

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #129645



cc @alexsamardzic @nikitaved @cpuhrsch @amjames @bhosmer @jcaip

pearu committed:
Add pinned memory support to sparse COO/CSR/CSC/BSR/BSC tensors

[ghstack-poisoned]

pearu committed:
Update on "Add pinned memory support to sparse COO/CSR/CSC/BSR/BSC tensors"

As in the title.

Fixes https://github.com/pytorch/pytorch/issues/115330




[ghstack-poisoned]

pearu committed:
Update on "Add pinned memory support to sparse COO/CSR/CSC/BSR/BSC tensors"

As in the title:

To register indices/values of a sparse XYZ tensor with CUDA, the following methods are supported
- `sparse_xyz_tensor(indices, values, pin_memory=True)`
- `sparse_xyz_tensor(indices, values).pin_memory()`
- `sparse_xyz_tensor(indices.pin_memory(), values.pin_memory())`

Fixes https://github.com/pytorch/pytorch/issues/115330




[ghstack-poisoned]

pearu committed:
Update on "Add pinned memory support to sparse COO/CSR/CSC/BSR/BSC tensors"

As in the title:

To register indices/values of a sparse XYZ tensor with CUDA, the following methods are supported
- `sparse_xyz_tensor(indices, values, pin_memory=True)`
- `sparse_xyz_tensor(indices, values).pin_memory()`
- `sparse_xyz_tensor(indices.pin_memory(), values.pin_memory())`

Fixes https://github.com/pytorch/pytorch/issues/115330




[ghstack-poisoned]

pearu committed:
Update on "Add pinned memory support to sparse COO/CSR/CSC/BSR/BSC tensors"

As in the title:

To register indices/values of a sparse XYZ tensor with CUDA, the following methods are supported
- `sparse_xyz_tensor(indices, values, pin_memory=True)`
- `sparse_xyz_tensor(indices, values).pin_memory()`
- `sparse_xyz_tensor(indices.pin_memory(), values.pin_memory())`

Fixes https://github.com/pytorch/pytorch/issues/115330




[ghstack-poisoned]

pearu committed:
Update on "Add pinned memory support to sparse COO/CSR/CSC/BSR/BSC tensors"

As in the title:

To register indices/values of a sparse XYZ tensor with CUDA, the following methods are supported
- `sparse_xyz_tensor(indices, values, pin_memory=True)`
- `sparse_xyz_tensor(indices, values).pin_memory()`
- `sparse_xyz_tensor(indices.pin_memory(), values.pin_memory())`

Fixes https://github.com/pytorch/pytorch/issues/115330




[ghstack-poisoned]

pearu committed:
Update on "Add pinned memory support to sparse COO/CSR/CSC/BSR/BSC tensors"

As in the title:

To register indices/values of a sparse XYZ tensor with CUDA, the following methods are supported
- `sparse_xyz_tensor(indices, values, pin_memory=True)`
- `sparse_xyz_tensor(indices, values).pin_memory()`
- `sparse_xyz_tensor(indices.pin_memory(), values.pin_memory())`

Fixes https://github.com/pytorch/pytorch/issues/115330




[ghstack-poisoned]

pearu committed:
Update on "Add pinned memory support to sparse COO/CSR/CSC/BSR/BSC tensors"

As in the title:

To register indices/values of a sparse XYZ tensor with CUDA, the following methods are supported
- `sparse_xyz_tensor(indices, values, pin_memory=True)`
- `sparse_xyz_tensor(indices, values).pin_memory()`
- `sparse_xyz_tensor(indices.pin_memory(), values.pin_memory())`

Fixes https://github.com/pytorch/pytorch/issues/115330




[ghstack-poisoned]

pearu committed:
Update on "Add pinned memory support to sparse COO/CSR/CSC/BSR/BSC tensors"

As in the title:

To register indices/values of a sparse XYZ tensor with CUDA, the following methods are supported
- `sparse_xyz_tensor(indices, values, pin_memory=True)`
- `sparse_xyz_tensor(indices, values).pin_memory()`
- `sparse_xyz_tensor(indices.pin_memory(), values.pin_memory())`

Fixes https://github.com/pytorch/pytorch/issues/115330




[ghstack-poisoned]

pearu committed:
Update on "Add pinned memory support to sparse COO/CSR/CSC/BSR/BSC tensors"

As in the title:

To register indices/values of a sparse XYZ tensor with CUDA, the following methods are supported
- `sparse_xyz_tensor(indices, values, pin_memory=True)`
- `sparse_xyz_tensor(indices, values).pin_memory()`
- `sparse_xyz_tensor(indices.pin_memory(), values.pin_memory())`

Fixes https://github.com/pytorch/pytorch/issues/115330




[ghstack-poisoned]

amjames reviewed (changes requested):
I see one major issue that should be addressed. 

What should happen when one component tensor passed to _with_tensors` constructors, but not all. In particular consider when the `values` are pinned, but indices tensor(s) are not.

```python
a = torch.randn(10, 10)
a[::2, ::2] = 0
sp1 = a.to_sparse_coo()
values = a._values().clone().pin_memory()
indices = a._indices()
sp2 = torch.sparse_coo_tensor(indices, values, size=(10, 10))
```

Here `sp2.is_pinned()` would return true, but `sp2._indices().is_pinned()` is false. This is inconsistent with

```python
sp3 = sp1.pin_memory()
```

which also has `sp3.is_pinned()` True, but `sp3._indices().is_pinned()` is True. Inconsistency like this will be a pain point for some users. 

I see two options: 
1. Expect this meta data field to match across members -> We should add this to the invariant checks, or enforce this by pinning out-of-sync members if any member is pinned in the constructors. 
2. Allow for these to be out-of-sync 
   - `sparse.is_pinned()` should be a short hand for `values.is_pinned() && indices.is_pinned()` (and the plain/compressed indices for sparse compressed layouts) the same way that `sparse.pin_memory()` is short hand for pinning all member tensors. 

Personally I think option 1) is better for the simplicity and I can't see a reason why one would want some members to be pinned while others are not.

pearu commented:
> I see two options: ...

Registering a CPU tensor with CUDA is essentially a copy operation to another device (although the pinned tensor can still be accessed by CPU processes). In the case of sparse tensors, we don't allow indices and values to be on different devices: an error is raised when trying to create a sparse tensor from indices and values that have different devices.

I suggest that we treat pinned and non-pinned tensors as having different devices. So, when one tries to create a sparse tensor from indices and values that have different pinning state then an exception will be raised. This corresponds to the first part of option 1).

pearu committed:
Update on "Add pinned memory support to sparse COO/CSR/CSC/BSR/BSC tensors"

As in the title:

To register indices/values of a sparse XYZ tensor with CUDA, the following methods are supported
- `sparse_xyz_tensor(indices, values, pin_memory=True)`
- `sparse_xyz_tensor(indices, values).pin_memory()`
- `sparse_xyz_tensor(indices.pin_memory(), values.pin_memory())`

Fixes https://github.com/pytorch/pytorch/issues/115330




cc alexsamardzic nikitaved cpuhrsch amjames bhosmer jcaip

[ghstack-poisoned]

pearu commented on aten/src/ATen/native/sparse/SparseTensor.cpp:
```diff
@@ -846,4 +851,22 @@ Tensor empty_like_sparse_coo(
+
+Tensor _pin_memory_sparse_coo(const Tensor& self, std::optional<Device> device) {
+  TORCH_INTERNAL_ASSERT_DEBUG_ONLY(!device.has_value() || device->is_cuda());
+  TensorOptions options = self.options().pinned_memory(true);
```
Done.

pearu commented on aten/src/ATen/native/sparse/SparseCsrTensor.cpp:
```diff
@@ -1219,4 +1232,25 @@ Tensor select_copy_sparse_csr(const Tensor& self, int64_t dim, int64_t index) {
+
+Tensor _pin_memory_sparse_compressed(const Tensor& self, std::optional<Device> device) {
+  TORCH_INTERNAL_ASSERT_DEBUG_ONLY(!device.has_value() || device->is_cuda());
+  CheckSparseTensorInvariants _(false);
```
Done.

pearu committed:
Update on "Add pinned memory support to sparse COO/CSR/CSC/BSR/BSC tensors"

As in the title:

To register indices/values of a sparse XYZ tensor with CUDA, the following methods are supported
- `sparse_xyz_tensor(indices, values, pin_memory=True)`
- `sparse_xyz_tensor(indices, values).pin_memory()`
- `sparse_xyz_tensor(indices.pin_memory(), values.pin_memory())`

Fixes https://github.com/pytorch/pytorch/issues/115330




cc alexsamardzic nikitaved cpuhrsch amjames bhosmer jcaip

[ghstack-poisoned]

pearu committed:
Update on "Add pinned memory support to sparse COO/CSR/CSC/BSR/BSC tensors"

As in the title:

To register indices/values of a sparse XYZ tensor with CUDA, the following methods are supported
- `sparse_xyz_tensor(indices, values, pin_memory=True)`
- `sparse_xyz_tensor(indices, values).pin_memory()`
- `sparse_xyz_tensor(indices.pin_memory(), values.pin_memory())`

Fixes https://github.com/pytorch/pytorch/issues/115330




cc alexsamardzic nikitaved cpuhrsch amjames bhosmer jcaip

[ghstack-poisoned]

pearu committed:
Update on "Add pinned memory support to sparse COO/CSR/CSC/BSR/BSC tensors"

As in the title:

To register indices/values of a sparse XYZ tensor with CUDA, the following methods are supported
- `sparse_xyz_tensor(indices, values, pin_memory=True)`
- `sparse_xyz_tensor(indices, values).pin_memory()`
- `sparse_xyz_tensor(indices.pin_memory(), values.pin_memory())`

Fixes https://github.com/pytorch/pytorch/issues/115330




cc alexsamardzic nikitaved cpuhrsch amjames bhosmer jcaip

[ghstack-poisoned]

pearu commented:
@pytorchbot merge
